### PR TITLE
add channels to juju charm list-resources and propagate it down to csclient.Client

### PR DESCRIPTION
**Pull request by natefinch**
(reviews: http://reviews.vapour.ws/r/4232/)

### CODE REVIEW

File: apiserver/charmrevisionupdater/updater.go [hunk 1]
```diff
@@ -145,7 +146,7 @@ func retrieveLatestCharmInfo(st *state.State) ([]latestCharmInfo, error) {
 		return nil, err
 	}
 
-	results, err := charmstore.LatestCharmInfo(client, curls)
+	results, err := charmstore.LatestCharmInfo(client, curls, "stable")
```
Comment (ericsnowcurrently) [inline apiserver/charmrevisionupdater/updater.go]: Unfortunately, there will be a mix of channels.  So you'll need to change charmstore.LatestCharmInfo to take a []charm.ChannelURL (or equivalent) where each item has the channel hard-coded to "stable".

File: charmstore/client.go [hunk 2]
```diff
@@ -35,8 +36,9 @@ type BaseClient interface {
 	// charmrepo/csclient/params, but we don't have one yet.
 
 	// LatestRevisions returns the latest revision for each of the
-	// identified charms. The revisions in the provided URLs are ignored.
-	LatestRevisions([]*charm.URL) ([]charmrepo.CharmRevision, error)
+	// identified charms in the given channel. The revisions in the provided
+	// URLs are ignored.
+	LatestRevisions(urls []*charm.URL, channel string) ([]charmrepo.CharmRevision, error)
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: As mentioned, the channels are not homogeneous.  So the channel must be associated with each element.

File: charmstore/client.go [hunk 3]
```diff
@@ -59,7 +61,7 @@ type ResourcesClient interface {
 	// list of details for each of the charm's resources. Those details
 	// are those associated with the specific charm revision. They
 	// include the resource's metadata and revision.
-	ListResources(charmURLs []*charm.URL) ([][]charmresource.Resource, error)
+	ListResources(charmURLs []*charm.URL, channel string) ([][]charmresource.Resource, error)
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: Likewise.

File: charmstore/client.go [hunk 4]
```diff
@@ -83,23 +82,74 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
...
+func (base baseClient) LatestRevisions(cURLs []*charm.URL, channel string) ([]charmrepo.CharmRevision, error) {
 	// TODO(ericsnow) Fix this:
 	// We must use charmrepo.CharmStore since csclient.Client does not
 	// have the "Latest" method.
-	repo := base.asRepo()
+
+	// unfortunately, the way the csclient.Client uses channels is pretty
+	// awkward for our uses.  It's an unexported field that only gets set by
+	// WithChannel, so we have to do this:
+	repo := charmrepo.NewCharmStoreFromClient(base.Client.WithChannel(params.Channel(channel)))
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: This doesn't seem quite right.

File: charmstore/client.go [hunk 5]
```diff
@@ -83,23 +82,74 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
...
 	return repo.Latest(cURLs...)
 }
 
+// GetResource implements BaseClient by calling csclient.Client's GetResource function.
+func (base baseClient) GetResource(cURL *charm.URL, resourceName string, revision int) (charmresource.Resource, io.ReadCloser, error) {
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: This needs the channel.

Comment (ericsnowcurrently) [inline charmstore/client.go]: oh, nope

File: charmstore/client.go [hunk 6]
```diff
@@ -83,23 +82,74 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
...
+	if err != nil {
+		return charmresource.Resource{}, nil, errors.Trace(err)
+	}
+	// TODO(natefinch): We don't actually have the full metadata here... how do
+	// we fix that?
+	return charmresource.Resource{
+		Meta: charmresource.Meta{
+			Name: resourceName,
+		},
+		Revision:    data.Revision,
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: You also need Origin (and the meta fields).

File: charmstore/client.go [hunk 7]
```diff
@@ -83,23 +82,74 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
...
+		Fingerprint: fp,
+		Size:        data.Size,
+	}, data.ReadCloser, nil
+}
+
+// ListResources implements BaseClient by calling csclient.Client's ListResources function.
+func (base baseClient) ListResources(charmURLs []*charm.URL, channel string) ([][]charmresource.Resource, error) {
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: Like the others, channel should be passed with each item.

File: charmstore/client.go [hunk 8]
```diff
@@ -184,29 +234,60 @@ func (c Client) Metadata() JujuMetadata {
 // identified charms. The revisions in the provided URLs are ignored.
 // Note that this differs from BaseClient.LatestRevisions() exclusively
 // due to taking into account Juju metadata (if any).
-func (c *Client) LatestRevisions(cURLs []*charm.URL) ([]charmrepo.CharmRevision, error) {
+func (c *Client) LatestRevisions(cURLs []*charm.URL, channel string) ([]charmrepo.CharmRevision, error) {
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: Again, channel-per-item.

File: charmstore/client.go [hunk 9]
```diff
@@ -83,23 +82,74 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
 	base := &baseClient{
 		Client: raw,
 	}
-	base.asRepo = func() *charmrepo.CharmStore {
```
Comment (ericsnowcurrently) [inline charmstore/client.go]: I'd prefer that this be left as-is.  It's a tricky bit.

File: charmstore/client.go (hunk 5 above)
Comment (natefinch) [inline charmstore/client.go]: yeah, revision is independent of channel, thankfully.

File: charmstore/client.go [hunk 10]
```diff
@@ -83,21 +82,111 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
...
 // LatestRevisions implements BaseClient.
-func (base baseClient) LatestRevisions(cURLs []*charm.URL) ([]charmrepo.CharmRevision, error) {
+func (base baseClient) LatestRevisions(charms []CharmID) ([]charmrepo.CharmRevision, error) {
 	// TODO(ericsnow) Fix this:
 	// We must use charmrepo.CharmStore since csclient.Client does not
 	// have the "Latest" method.
-	repo := base.asRepo()
-	return repo.Latest(cURLs...)
+
+	// unfortunately, the way the csclient.Client uses channels is pretty
```
Comment (rogpeppe) [inline charmstore/client.go]: s/unfortunately/Unfortunately/
Although perhaps just lose the word - it doesn't add much.

File: charmstore/client.go [hunk 11]
```diff
@@ -83,21 +82,111 @@ func newBaseClient(raw *csclient.Client, config ClientConfig, meta JujuMetadata)
...
+	// WithChannel, so we have to do make one bulk request per channel.
+
+	// collate charms into map[channel]charmIndex
+	req := collate(charms)
+
+	results := make([]charmrepo.CharmRevision, len(charms))
+
+	for channel, ci := range req {
+		repo := charmrepo.NewCharmStoreFromClient(base.Client.WithChannel(params.Channel(channel)))
+		revisions, err := repo.Latest(ci.ids...)
```
Comment (rogpeppe) [inline charmstore/client.go]: // TODO make these requests concurrently.
?

Comment (rogpeppe) [inline charmstore/client.go]: Or maybe it really doesn't matter in fact, because this isn't used often.

**Comment by natefinch** [comment]
This PR is being rebased against master, since feature-resources has landed permanently.

**Pull request closed without merging**